netquery/manager.go

```diff
@@ -0,0 +1,225 @@
+	// TODO(ppacher):
+	//		we don't yet support querying using the runtime database here ...
+	//		consider exposing connection from the database at least by ID.
+	//
+	// NOTE(ppacher):
+	//		for debugging purposes use RuntimeQueryRunner to execute plain
+	//		SQL queries against the database using portbase/database/runtime.
+	return nil, nil
+}
+
+// HandleFeed starts reading new and updated connections from feed and persists them
+// in the configured ConnectionStore. HandleFeed blocks until either ctx is cancelled
+// or feed is closed.
+// Any errors encountered when processing new or updated connections are logged but
+// otherwise ignored.
+// HandleFeed handles and persists updates one after each other! Depending on the system
+// load the user might want to use a buffered channel for feed.
+func (mng *Manager) HandleFeed(ctx context.Context, feed <-chan *network.Connection) {
+	// count the number of inserted rows for logging purposes.
+	//
+	// TODO(ppacher): how to handle the, though unlikely case, of a
+	// overflow to 0 here?
+	var count uint64
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+
+		case conn, ok := <-feed:
+			if !ok {
+				return
+			}
+
+			model, err := convertConnection(conn)
+			if err != nil {
+				log.Errorf("netquery: failed to convert connection %s to sqlite model: %w", conn.ID, err)
+
+				continue
+			}
+
+			log.Infof("netquery: persisting create/update to connection %s", conn.ID)
+
+			if err := mng.store.Save(ctx, *model); err != nil {
+				log.Errorf("netquery: failed to save connection %s in sqlite database: %w", conn.ID, err)
+
+				continue
+			}
+
+			// push an update for the connection
+			if err := mng.pushConnUpdate(ctx, *model); err != nil {
+				log.Errorf("netquery: failed to push update for conn %s via database system: %w", conn.ID, err)
+			}
+
+			count++
+
+			if count%20 == 0 {
+				log.Debugf("netquery: persisted %d connections so far", count)
+			}
+		}
+	}
+}
+
+func (mng *Manager) pushConnUpdate(ctx context.Context, conn Conn) error {
+	blob, err := json.Marshal(conn)
+	if err != nil {
+		return fmt.Errorf("failed to marshal connection: %w", err)
+	}
+
+	key := fmt.Sprintf("%s:%s%s", mng.runtimeReg.DatabaseName(), mng.pushPrefix, conn.ID)
+	wrapper, err := record.NewWrapper(
+		key,
+		new(record.Meta),
+		dsd.JSON,
+		blob,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create record wrapper: %w", err)
+	}
+
+	// FIXME(ppacher): it may happen that started != now for NEW connections.
+	// In that case we would push and UPD rather than NEW even if
+	// the connection is new ...
+	// Though, that's still better than always pushing NEW for existing
+	// connections.
+	// If we would use UnixNano() here chances would be even worse.
+	//
+	// Verify if the check in portbase/api/database.go is vulnerable
+	// to such timing issues in general.
+	wrapper.SetMeta(&record.Meta{
+		Created:  conn.Started.Unix(),
+		Modified: time.Now().Unix(),
+	})
```

Comment:
You can just set the data in the NewWrapper call above (and save one allocation).
Is there some way to get an indicator from SQLite whether the record was inserted or updated?

Reply:
Copy meta from connection.